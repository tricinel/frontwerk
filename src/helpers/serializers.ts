import slash from 'slash';

// Remove the quotes around strings
const unquoteSerializer = {
  print: <T>(val: T): T => val,
  test: <T>(val: T): boolean => typeof val === 'string'
};

// Cconvert windows style file paths to unix
const winPathSerializer = {
  print: (val: string): string => slash(val),
  test: <T>(val: T): boolean => typeof val === 'string' && val.includes('\\')
};

export { unquoteSerializer, winPathSerializer };
