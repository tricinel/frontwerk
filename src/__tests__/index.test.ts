/* eslint-disable global-require */
import jestSerializerPath from 'jest-serializer-path';
import cases from 'jest-in-case';

import { unquoteSerializer, winPathSerializer } from '../helpers/serializers';

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

jest.mock('cross-spawn');

const testCases = [
  {
    name: 'errors out when using the wrong node version',
    nodeVersion: 'v6.11.0',
    throws: true
  },
  {
    name: 'calls node with the script path and args',
    args: ['test', '--no-watch']
  },
  {
    name: 'throws unknown script',
    args: ['unknown-script'],
    throws: true
  },
  {
    name: 'logs help with no args',
    snapshotLog: true
  }
];

const testFn = ({
  snapshotLog = false,
  throws = false,
  args = [],
  nodeVersion = 'v8.4.0'
}) => {
  const { sync: crossSpawnSyncMock } = require('cross-spawn');
  const originalExit = jest.spyOn(process, 'exit');
  const originalLog = jest.spyOn(console, 'log');
  const { argv: originalArgv } = process;

  delete process.version;
  process.version = nodeVersion;

  try {
    // Tests
    process.argv = ['node', '../', ...args];
    crossSpawnSyncMock.mockClear();
    require('../');

    if (snapshotLog) {
      expect(originalLog.mock.calls).toMatchSnapshot();
    } else {
      expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1);
      const [firstCall] = crossSpawnSyncMock.mock.calls;
      const [script, calledArgs] = firstCall;
      expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
    }
  } catch (error) {
    if (throws) {
      expect(error.message).toMatchSnapshot();
    } else {
      throw error;
    }
  } finally {
    // We reset everything afterEach
    originalExit.mockRestore();
    originalLog.mockRestore();
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('format', testFn, testCases);
