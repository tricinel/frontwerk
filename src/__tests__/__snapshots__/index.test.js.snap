// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format calls node with the script path and args 1`] = `node <PROJECT_ROOT>/src/scripts/test.js --no-watch`;

exports[`format errors out when using the wrong node version 1`] = `You must use Node version 8 or greater`;

exports[`format logs help with no args 1`] = `
Array [
  Array [
    
Usage: ../ [script] [--flags]
Available Scripts:
build --bundle: Build your ES6 files into a single file using RollupJS
build: Build your ES6 files using BabelJS
lint: Lint your JS and Typescript files using ESLint and TSLint
tslint: Lint your Typescript files using TSLint
stylelint: Lint your CSS/SCSS/SASS files using Stylelint
prettier: REMOVED: Please use format instead
format: Format your JS files using Prettier
test: Test your JS files using Jest
flow: Run your files through the flow static type checker
eject: Tired of frontwerk? Eject will copy all the required configs and packages to your app and remove itself afterwards.
Options:
  All options depend on the script. The args you pass will be forwarded to the respective tool that's being run under the hood.
,
  ],
]
`;

exports[`format throws unknown script 1`] = `Unknown script unknown-script. Please check if you need to update frontwerk.`;
